torchvision/prototype/datapoints/_dataset_wrapper.py

```diff
@@ -0,0 +1,250 @@
+
+
+class _VisionDatasetDatapointWrapper(Dataset):
+    def __init__(self, dataset: datasets.VisionDataset, wrapper) -> None:
+        self.vision_dataset = dataset
+        self.wrapper = wrapper
+
+        # We need to disable the transforms on the dataset here to be able to inject the wrapping before we apply the
+        # transforms
+        self.transform, dataset.transform = dataset.transform, None
+        self.target_transform, dataset.target_transform = dataset.target_transform, None
+        self.transforms, dataset.transforms = dataset.transforms, None
+
+    def __getattr__(self, item: str) -> Any:
+        with contextlib.suppress(AttributeError):
+            return object.__getattribute__(self, item)
+
+        return getattr(self.vision_dataset, item)
+
+    def __getitem__(self, idx: int) -> Any:
+        # This gets us the raw sample since we disabled the transforms for the underlying dataset in the constructor
+        # of this class
+        sample = self.vision_dataset[idx]
+
+        sample = self.wrapper(self.vision_dataset, sample)
+
+        # We don't need to care about `transform` and `target_transform` here since `VisionDataset` joins them into a
+        # `transforms` internally:
+        # https://github.com/pytorch/vision/blob/2d92728341bbd3dc1e0f1e86c6a436049bbb3403/torchvision/datasets/vision.py#L52-L54
+        if self.transforms is not None:
+            sample = self.transforms(*sample)
+
+        return sample
+
+    def __len__(self) -> int:
+        return len(self.vision_dataset)
+
+
+@functools.lru_cache(maxsize=None)
+def get_categories(dataset: datasets.VisionDataset) -> Optional[List[str]]:
+    categories_fn = {
+        datasets.Caltech256: lambda dataset: [name.rsplit(".", 1)[1] for name in dataset.categories],
+        datasets.CIFAR10: lambda dataset: dataset.classes,
+        datasets.CIFAR100: lambda dataset: dataset.classes,
+        datasets.FashionMNIST: lambda dataset: dataset.classes,
+        datasets.ImageNet: lambda dataset: [", ".join(names) for names in dataset.classes],
+    }.get(type(dataset))
+    return categories_fn(dataset) if categories_fn is not None else None
+
+
+def classification_wrapper(
+    dataset: datasets.VisionDataset, sample: Tuple[PIL.Image.Image, Optional[int]]
+) -> Tuple[PIL.Image.Image, Optional[datapoints.Label]]:
+    image, label = sample
+    if label is not None:
+        label = datapoints.Label(label, categories=get_categories(dataset))
+    return image, label
+
+
+for dataset_type in [
+    datasets.Caltech256,
+    datasets.CIFAR10,
+    datasets.CIFAR100,
+    datasets.ImageNet,
+    datasets.MNIST,
+    datasets.FashionMNIST,
+]:
+    _WRAPPERS[dataset_type] = classification_wrapper
+
+
+def segmentation_wrapper(
+    dataset: datasets.VisionDataset, sample: Tuple[PIL.Image.Image, PIL.Image.Image]
+) -> Tuple[PIL.Image.Image, datapoints.Mask]:
+    image, mask = sample
+    return image, datapoints.Mask(F.to_image_tensor(mask))
```

Comment:
Curious why we need `F.to_image_tensor(mask)` before wrapping it into a mask. Does that mean `Image(some_pil_image)` doesn't work either?

Reply:
Yes, so far that doesn't work either. So far we pass the input directly into `torch.as_tensor` and PIL images don't work here. IMO, it was preferable to support that directly, or at least have something like `Image.from_pil` for convenience. I proposed something like that in the past (can't find the issue now), but it was soft-rejected since we have this functionality as transform already.

Reply:
I feel like this is precisely the kind of APIs we need to make as frictionless as possible if we want to optimize adoption / migration. If users have to make 2 calls to convert their PIL images into a DataPoint Image, that's one too many calls (IIUC the above you would agree?).

There might be good technical reasons of *why* this may be hard to enable, but this seems like a very natural and expected UX. Is this something we can revisit?

Reply:
> There might be good technical reasons of why this may be hard to enable

I don't think that is hard to enable. Looking at the constructor of `Image`, we have
https://github.com/pytorch/vision/blob/135a0f9ea9841b6324b4fe8974e2543cbb95709a/torchvision/prototype/datapoints/_image.py#L18-L26

I don't think it takes more than something like

```python
if isinstance(data, PIL.Image.Image):
    data = F.pil_to_tensor(data)
data = cls._to_tensor(...)
```

> Is this something we can revisit?

As stated above, I wanted to have it in the first place so no concerns from my side. @vfdev-5?